go-controller/pkg/clustermanager/pod/allocator.go

```diff
@@ -42,28 +50,39 @@ type PodAllocator struct {
 }
 
 // NewPodAllocator builds a new PodAllocator
-func NewPodAllocator(netInfo util.NetInfo, podLister listers.PodLister, kube kube.Interface) *PodAllocator {
-	podAnnotationAllocator := pod.NewPodAnnotationAllocator(
-		netInfo,
-		podLister,
-		kube,
-	)
-
+func NewPodAllocator(
+	netInfo util.NetInfo,
+	kube kube.InterfaceOVN,
+	podAnnotationAllocator pod.AnnotationAllocator,
+	ipAllocator subnet.Allocator,
+	opts ...AllocationOption,
+) *PodAllocator {
 	podAllocator := &PodAllocator{
 		netInfo:                netInfo,
 		releasedPods:           map[string]sets.Set[string]{},
 		releasedPodsMutex:      sync.Mutex{},
 		podAnnotationAllocator: podAnnotationAllocator,
+		kube:                   kube,
 	}
 
 	// this network might not have IPAM, we will just allocate MAC addresses
 	if util.DoesNetworkRequireIPAM(netInfo) {
-		podAllocator.ipAllocator = subnet.NewAllocator()
+		podAllocator.ipAllocator = ipAllocator
+	}
+
+	for _, opt := range opts {
+		opt(podAllocator)
 	}
 
 	return podAllocator
 }
 
+func WithPersistentIPs(ipamClaimsLister ipamclaimslister.IPAMClaimLister) AllocationOption {
+	return func(allocator *PodAllocator) {
+		allocator.ipamClaimsFetcher = persistentips.NewIPAMClaimsFetcher(allocator.netInfo, ipamClaimsLister)
+	}
+}
```

Comment:
> > I guess then you would need to check if `allocator.ipamClaimsFetcher` is `nil` or not somewhere?
> 
> OK, I can do it.
> 
> > Why wouldn't you pass the ipam claim fetcher that the user already has instead of building a new one?
> 
> No good reason. I can pass it along from the caller.
> 
> > Also this is using a different approach than how we handle IPAM and tunnel IDs variations.
> 
> > Wouldn't it make sense to refactor this to be consistent throughout this receiver so that it is easier to understand?. So we either have also WithIPAM and WithTunnelIDs options, or we handle ipam claims like everything else, querying the netinfo internally to decide whether we should use the fetcher that will be passed in the constructor
> 
> I see your point. I agree the code would look more consistent either of those options are taken.
> 
> I'll check if refactoring the tunnel part to use a builder is as simple as I think it is, and will follow through unless it proves too much. If so, I'll just inject everything via the constructor.
> 
> > Also wouldn't it make sense to handle the subnet allocator setup somewhere else given that this is no longer the owner. The owner should take care that it is setup correctly before passing it around I guess.
> 
> Could you elaborate a bit here ? Do you mean `ipAllocator.AddOrUpdateSubnet` should be done by the caller ?

Yeah, I guess that should be setup somewhere in the networkClusterController now.

Reply:
Done in [this commit](https://github.com/ovn-org/ovn-kubernetes/pull/4165/commits/742019d080443289f8d6bdf19519dea114b4fab4). 

I think it looks good, and can be squashed into the proper place.

Your thoughts ?